rskj-core/src/main/java/co/rsk/net/sync/DownloadingSkeletonSyncState.java

```diff
@@ -19,20 +36,58 @@ public boolean isSyncing() {
     }
 
     @Override
-    public void newSkeleton(List<BlockIdentifier> skeleton) {
+    public void newSkeleton(List<BlockIdentifier> skeleton, MessageChannel peer) {
+        NodeID peerId = peer.getPeerNodeID();
+        boolean isSelectedPeer = peerId == syncInformation.getSelectedPeerId();
+
         // defensive programming: this should never happen
         if (skeleton.size() < 2) {
-            syncEventsHandler.onErrorSyncing(
-                    "Invalid skeleton received from node {}",
-                    syncInformation.getSelectedPeerId());
-            return;
+            syncInformation.reportEvent("Invalid skeleton received from node {}",
+                    EventType.INVALID_MESSAGE, peerId);
+
+            // when the selected peer fails automatically all process restarts
+            if (isSelectedPeer){
+                syncEventsHandler.stopSyncing();
+                return;
+            }
+        } else {
+            skeletons.put(peerId, skeleton);
         }
 
-        syncEventsHandler.startDownloadingHeaders(skeleton, connectionPoint);
+        expectedSkeletons--;
+        selectedPeerAnswered = selectedPeerAnswered || isSelectedPeer;
+
+        if (expectedSkeletons <= 0){
+            if (skeletons.size() == 0){
+                syncEventsHandler.stopSyncing();
+                return;
+            }
+            syncEventsHandler.startDownloadingHeaders(skeletons, connectionPoint);
+        }
+    }
+
+    @Override
+    public void tick(Duration duration) {
+        timeElapsed = timeElapsed.plus(duration);
+        if (timeElapsed.compareTo(syncConfiguration.getTimeoutWaitingRequest()) >= 0) {
+            candidates.stream()
+                    .filter(c -> !skeletons.containsKey(c.getPeerNodeID()))
+                    .forEach(p ->
+                            syncInformation.reportEvent("Timeout waiting requests from node {}",
+                                    EventType.TIMEOUT_MESSAGE, p.getPeerNodeID()));
+
+            // when the selected peer fails automatically all process restarts
+            if (!selectedPeerAnswered){
+                syncEventsHandler.stopSyncing();
+                return;
+            }
+
+            syncEventsHandler.startDownloadingHeaders(skeletons, connectionPoint);
```

Comment:
[MAYBE BUG] Do we start downloading headers only when we timeout??

Reply:
No, also when a new skeleton arrives and we are not expecting any more skeletons to arrive we start to download the headers